### PR TITLE
Add colouring to wrong words

### PR DESCRIPTION
**Pull request by ailin-nemui**
Highlight:

wrong words can be highlighted in the input prompt:

spellcheck

### CODE REVIEW

File: spellcheck.pl [hunk 1]
```diff
@@ -16,9 +16,11 @@
...
 use vars qw($VERSION %IRSSI);
 use Irssi 20070804;
+use Irssi::TextUI;
+use Encode;
 use Text::Aspell;
 
-$VERSION = '0.8.1';
+$VERSION = '0.8.2';
 %IRSSI = (
     authors     => 'Jakub Wilk, Jakub Jankowski, Gabriel Pettier',
```
Comment (jwilk) [inline spellcheck.pl]: Please consider adding yourself to the list of authors (and to the list of copyright holders earlier in the file).

File: spellcheck.pl [hunk 2]
```diff
@@ -60,6 +62,9 @@ sub spellcheck_check_word
     }
 
     return if $word =~ m{^/}; # looks like a path
+    if (Irssi::settings_get_bool('expand_escapes')) {
+	$word =~ s/\\(['"\\])/$1/g;
```
Comment (jwilk) [inline spellcheck.pl]: I don't understand why is this needed.
Please use spaces, not tabs.

File: spellcheck.pl [hunk 3]
```diff
@@ -135,30 +140,39 @@ sub spellcheck_key_pressed
...
     return unless Irssi::settings_get_bool('spellcheck_enabled');
 
-    # I know no way to *mark* misspelled words in the input line,
-    # that's why there's no spellcheck_print_suggestions -
-    # because printing suggestions is our only choice.
-
     # hide correction window when message is sent
     if (chr($key) =~ /\A[\r\n]\z/ && $correction_window) {
         $correction_window->command("window hide $window_name");
+        if (Irssi->can('gui_input_clear_extents')) {
```
Comment (jwilk) [inline spellcheck.pl:143]: This API doesn't seem to be present in the latest Irssi release (1.1.1). Is that correct?
In any case, minimum version requirements should be documented in the commit messages, and eventually also in README.

File: spellcheck.pl [hunk 4]
```diff
@@ -135,30 +140,39 @@ sub spellcheck_key_pressed
...
+            Irssi::gui_input_clear_extents(0, 9999);
+        }
+    }
+
+    # get current inputline
+    my $inputline = Irssi::parse_special('$L');
+    if (Irssi::settings_get_str('term_charset') =~ /utf-?8/i) {
```
Comment (jwilk) [inline spellcheck.pl]: The regexp should be anchored (with `\A` and `\z`)…

File: spellcheck.pl [hunk 5]
```diff
@@ -135,30 +140,39 @@ sub spellcheck_key_pressed
...
+        Encode::_utf8_on($inputline);
```
Comment (jwilk) [inline spellcheck.pl:151]: …but maybe it would be better to just decode the input line using `term_charset` encoding? Then you wouldn't need any ifs or regexps.

File: spellcheck.pl [hunk 6]
```diff
@@ -135,30 +140,39 @@ sub spellcheck_key_pressed
...
+    }
+    if (Irssi->can('gui_input_set_extent')) {
+	Irssi::gui_input_set_extent(length $inputline, '%n');
```
Comment (jwilk) [inline spellcheck.pl]: tabs → spaces

File: spellcheck.pl [hunk 7]
```diff
@@ -135,30 +140,39 @@ sub spellcheck_key_pressed
...
-    # get current inputline
-    my $inputline = Irssi::parse_special('$L');
+    $inputline = substr $inputline, 0, Irssi::gui_input_get_pos;
 
     # check if inputline starts with any of cmdchars
     # we shouldn't spell-check commands
-    my $cmdchars = Irssi::settings_get_str('cmdchars');
-    my $re = qr/^[$cmdchars]/;
-    return if ($inputline =~ $re);
+    return unless -1 == index(Irssi::settings_get_str('cmdchars'), substr($inputline, 0, 1));
```
Comment (jwilk) [inline spellcheck.pl]: I guess the original code will break if `cmdchars` contains `\` or `]`.
Is that the only bug you're fixing here?
Anyway, this seems unrelated to coloring words on input line, so it should be in a separate commit.

File: spellcheck.pl [hunk 8]
```diff
@@ -135,30 +140,39 @@ sub spellcheck_key_pressed
...
 
     # get last bit from the inputline
-    my ($word) = $inputline =~ /\s*(\S+)$/;
+    my ($word) = $inputline =~ /\s*(\S+)[.\s]*$/;
```
Comment (jwilk) [inline spellcheck.pl:169]: Why?

File: spellcheck.pl [hunk 9]
```diff
@@ -168,10 +182,18 @@ sub spellcheck_key_pressed
...
+    }
+
+    return unless Irssi::settings_get_bool('spellcheck_print_suggestions');
+
     # show corrections window if hidden
     if ($correction_window) {
-        $win->command("window show $window_name");
-        $correction_window->command('window stick off');
+        $win->command("^window show $window_name");
+        $correction_window->command('^window stick off');
```
Comment (jwilk) [inline spellcheck.pl]: What does leading `^` do?

File: spellcheck.pl [hunk 10]
```diff
@@ -168,10 +182,18 @@ sub spellcheck_key_pressed
...
+        $win->set_active;
```
Comment (jwilk) [inline spellcheck.pl]: Why?

File: spellcheck.pl [hunk 11]
```diff
@@ -239,13 +261,17 @@ sub spellcheck_add_word
...
 Irssi::settings_add_str( 'spellcheck', 'spellcheck_languages', '');
 Irssi::settings_add_str( 'spellcheck', 'spellcheck_word_color', '%R');
 Irssi::settings_add_str( 'spellcheck', 'spellcheck_window_name', '');
 Irssi::settings_add_str( 'spellcheck', 'spellcheck_window_height', 10);
+Irssi::settings_add_str( 'spellcheck', 'spellcheck_word_input_color', '%U');
 
-Irssi::signal_add_first('gui key pressed', 'spellcheck_key_pressed');
+Irssi::signal_add_last('key word_completion', sub{spellcheck_key_pressed(ord '.')});
+Irssi::signal_add_last('key word_completion_backward', sub{spellcheck_key_pressed(ord '.')});
+Irssi::signal_add_last('gui key pressed', 'spellcheck_key_pressed');
```
Comment (jwilk) [inline spellcheck.pl:273]: What do the changes to `signal_add_…` do?

**Review by jwilk — CHANGES_REQUESTED** [review]
I haven't tested this yet, but the screenshot looks awesome.

The author email for this commit is `ailin@z30a.localdomain`, which isn't associated with your GitHub account and doesn't look correct all.

File: spellcheck.pl (hunk 11 above)
Comment (ailin-nemui) [reply spellcheck.pl:273]: the highlight should be updated when you use tab (word_completion), and it should also encompass the last typed letter (thus last instead of first)

File: spellcheck.pl (hunk 9 above)
Comment (ailin-nemui) [reply spellcheck.pl]: hide messages like window is not sticky/window is already visible

File: spellcheck.pl (hunk 8 above)
Comment (ailin-nemui) [reply spellcheck.pl:169]: the input line will now end with what we typed, so we should ignore/allow space and .

File: spellcheck.pl (hunk 7 above)
Comment (ailin-nemui) [reply spellcheck.pl]: you're right I fixed a bug here

File: spellcheck.pl (hunk 3 above)
Comment (ailin-nemui) [reply spellcheck.pl:143]: try irssi from git / 1.2

File: spellcheck.pl (hunk 2 above)
Comment (ailin-nemui) [reply spellcheck.pl]: it's another bug I fixed, if this setting is enabled we should replace \\'  with  '  when spell checking words such as "it's"

File: spellcheck.pl (hunk 10 above)
Comment (ailin-nemui) [reply spellcheck.pl]: another bug fix, when we run /win show, the input focus will change to the "spell" window. but we do not want this, so set the focus back to the original window

**Comment by ailin-nemui** [comment]
I attempted to implement your suggestions.

File: spellcheck.pl (hunk 7 above)
Comment (jwilk) [reply spellcheck.pl]: I commited a different fix for this: 62b1fdb561ff5387c2374170ff7745b369eadb2b
Thanks for the bug report.

File: spellcheck.pl [hunk 12]
```diff
@@ -60,6 +60,9 @@ sub spellcheck_check_word
     }
 
     return if $word =~ m{^/}; # looks like a path
+    if (Irssi::settings_get_bool('expand_escapes')) {
+        $word =~ s/\\(['"\\])/$1/g;
```
Comment (jwilk) [inline spellcheck.pl]: I don't think that's right.
My Irssi (1.0.7) doesn't expand `\'` to `'` even when `expand_escapes` is enabled.
I've filed separate bug (https://github.com/jwilk/irssi-spellcheck/issues/2) for proper support for `expand_escapes`.

Comment (ailin-nemui) [reply spellcheck.pl]: I see, we're both right. This was due to the fix that landed in Irssi 1.1.0 https://github.com/irssi/irssi/pull/709 but I think it won't hurt on earlier versions

File: spellcheck.pl (hunk 5 above)
Comment (ailin-nemui) [reply spellcheck.pl:151]: I don't think that's economic

File: spellcheck.pl (hunk 12 above)
Comment (ailin-nemui) [reply spellcheck.pl]: I dropped this patch as it doesn't make sense and the Irssi feature has a bug

**Comment by jwilk** [comment]
Can I have your full name in git commits metadata and in the `%IRSSI` hash? (I accept "no" as an anwer.)

File: spellcheck.pl [hunk 13]
```diff
@@ -239,13 +260,17 @@ sub spellcheck_add_word
 Irssi::command_bind('spellcheck_add', 'spellcheck_add_word');
 
 Irssi::settings_add_bool('spellcheck', 'spellcheck_enabled', 1);
+Irssi::settings_add_bool('spellcheck', 'spellcheck_print_suggestions', 1);
 Irssi::settings_add_str( 'spellcheck', 'spellcheck_default_language', 'en_US');
 Irssi::settings_add_str( 'spellcheck', 'spellcheck_languages', '');
 Irssi::settings_add_str( 'spellcheck', 'spellcheck_word_color', '%R');
 Irssi::settings_add_str( 'spellcheck', 'spellcheck_window_name', '');
 Irssi::settings_add_str( 'spellcheck', 'spellcheck_window_height', 10);
+Irssi::settings_add_str( 'spellcheck', 'spellcheck_word_input_color', '%U');
```
Comment (jwilk) [inline spellcheck.pl:269]: Any particular reason the default is underline, rather than red? I think on the screenshot red looked pretty good.

Comment (ailin-nemui) [reply spellcheck.pl:269]: just what I liked

**Comment by ailin-nemui** [comment]
I have several scripts published with my nick as author so I'd leave it

File: spellcheck.pl [hunk 14]
```diff
@@ -170,8 +170,9 @@ sub spellcheck_key_pressed
 
     # show corrections window if hidden
     if ($correction_window) {
-        $win->command("window show $window_name");
-        $correction_window->command('window stick off');
+        $win->command("^window show $window_name");
+        $correction_window->command('^window stick off');
```
Comment (jwilk) [inline spellcheck.pl]: I fixed another similar issue and commited it as 7c06d8c507c1824896d13cc2a2914efde1c7fdb9.

File: spellcheck.pl [hunk 15]
```diff
@@ -170,8 +170,9 @@ sub spellcheck_key_pressed
...
+        $win->set_active;
```
Comment (jwilk) [inline spellcheck.pl]: Cherry-picked as 4efbb19f0a75b306eedb0925475fed552bda936c.

File: spellcheck.pl [hunk 16]
```diff
@@ -16,11 +16,13 @@
 
 use vars qw($VERSION %IRSSI);
 use Irssi 20070804;
+use Irssi::TextUI;
+use Encode;
 use Text::Aspell;
 
-$VERSION = '0.8.1';
+$VERSION = '0.8.2';
```
Comment (jwilk) [inline spellcheck.pl]: 0.8.1 hasn't been released yet, so there was no need to bump this.

File: spellcheck.pl [hunk 17]
```diff
@@ -135,20 +137,27 @@ sub spellcheck_key_pressed
...
+        }
+    }
+
+    # get current inputline
+    my $inputline = Irssi::parse_special('$L');
+    if (lc Irssi::settings_get_str('term_charset') eq 'utf-8') {
+        Encode::_utf8_on($inputline);
+    }
+    if (Irssi->can('gui_input_set_extent')) {
+        Irssi::gui_input_set_extent(length $inputline, '%n');
```
Comment (jwilk) [inline spellcheck.pl:154]: What was this supposed to do?
It seems to make sure that when you correct an already-colored word, the newly added letters are not colored. But this only works at the end of the input line.

File: spellcheck.pl (hunk 8 above)
Comment (jwilk) [reply spellcheck.pl:169]: Ignoring the dot doesn't work, because it will be consumed by `\S+`. But meh, we can fix it later.

**Comment by jwilk** [comment]
> I have several scripts published with my nick as author

Fair enough, but please be consistent. f94e80fb14d11342e71307d11120dcf4a385c039 has a different author than ca448ca45cb7d2c6dd402f16f2d67fe92a9074d8.

File: spellcheck.pl (hunk 17 above)
Comment (ailin-nemui) [reply spellcheck.pl:154]: good point. you need to ensure that the line ends with %n always, otherwise if you delete the last character the underline will go to the end of the input prompt. but in fact the same issue will happen if you delete the last character of an underlined word :thinking:

**Comment by jwilk** [comment]
Merged as eb33fe308686cbd0e062c144d1719a43115be714 and 281d7ea1cb3e14fb7bef409ee95d91fdde53ea9a.

There are still some rough edges, but we can fix them later.

Thanks for implementing this excellent feature and sorry for taking so long to merge it!
